### DISCUSSION TITLE
Borg Backup Error During Prune

### DISCUSSION THREAD

**Discussion by DrShivang** [nextcloud/all-in-one#714] [Questions] [answered]
Hi, After the last update am getting this error during the backup and backup doesn't completes/succeeds.

Borg Error Log As Below 

> Pruning the archives...
Replaying segments   0%
g segments   5%
g segments  10%
g segments  15%
g segments  20%
g segments  25%
g segments  30%
g segments  35%
g segments  40%
g segments  45%
g segments  50%
g segments  55%
g segments  60%
g segments  65%
g segments  70%
g segments  75%
g segments  80%
g segments  85%
stale exclusive roster lock for host f6b5173ea560@2485377892355 pid 26 thread 0.
Removed stale exclusive roster lock for host f6b5173ea560@2485377892355 pid 26 thread 0.
Exception ignored in: <function Repository.__del__ at 0x7f9e8f8790>
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/borg/repository.py", line 179, in __del__
    self.close()
  File "/usr/lib/python3/dist-packages/borg/repository.py", line 460, in close
    self.lock.release()
  File "/usr/lib/python3/dist-packages/borg/locking.py", line 383, in release
    self._roster.modify(EXCLUSIVE, REMOVE)
  File "/usr/lib/python3/dist-packages/borg/locking.py", line 282, in modify
    elements.remove(self.id)
KeyError: ('f6b5173ea560@2485377892355', 29, 0)
Local Exception
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 164, in wrapper
    kwargs['manifest'], kwargs['key'] = Manifest.load(repository, compatibility)
  File "/usr/lib/python3/dist-packages/borg/helpers.py", line 378, in load
    cdata = repository.get(cls.MANIFEST_ID)
  File "/usr/lib/python3/dist-packages/borg/repository.py", line 1122, in get
    self.index = self.open_index(self.get_transaction_id())
  File "/usr/lib/python3/dist-packages/borg/repository.py", line 401, in get_transaction_id
    self.check_transaction()
  File "/usr/lib/python3/dist-packages/borg/repository.py", line 398, in check_transaction
    self.replay_segments(replay_from, segments_transaction_id)
  File "/usr/lib/python3/dist-packages/borg/repository.py", line 858, in replay_segments
    self._update_index(segment, objects)
  File "/usr/lib/python3/dist-packages/borg/repository.py", line 868, in _update_index
    for tag, key, offset, size in objects:
  File "/usr/lib/python3/dist-packages/borg/repository.py", line 1435, in iter_objects
    size, tag, key, data = self._read(fd, self.header_fmt, header, segment, offset,
  File "/usr/lib/python3/dist-packages/borg/repository.py", line 1532, in _read
    **raise IntegrityError('Segment entry checksum mismatch [segment {}, offset {}]'.format(
borg.helpers.IntegrityError: Data integrity error: Segment entry checksum mismatch [segment 59, offset 58849132]
During handling of the above exception, another exception occurred:**
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 4690, in main
    exit_code = archiver.run(args)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 4622, in run
    return set_ec(func(args))
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 177, in wrapper
    return method(self, args, repository=repository, **kwargs)
  File "/usr/lib/python3/dist-packages/borg/repository.py", line 204, in __exit__
    self.close()
  File "/usr/lib/python3/dist-packages/borg/repository.py", line 460, in close
    self.lock.release()
  File "/usr/lib/python3/dist-packages/borg/locking.py", line 383, in release
    self._roster.modify(EXCLUSIVE, REMOVE)
  File "/usr/lib/python3/dist-packages/borg/locking.py", line 282, in modify
    elements.remove(self.id)
KeyError: ('f6b5173ea560@2485377892355', 29, 0)
Platform: Linux f6b5173ea560 5.10.103-v8 #1 SMP PREEMPT Mon Mar 28 11:44:31 UTC 2022 aarch64
Linux: Unknown Linux  
Borg: 1.1.16  Python: CPython 3.9.2 msgpack: 0.5.6.+borg1
PID: 29  CWD: /
sys.argv: ['/usr/bin/borg', 'prune', '--prefix', '*_*-nextcloud-aio', '--stats', '--progress', '--keep-within=7d', '--keep-weekly=4', '--keep-monthly=6', '/mnt/borgbackup/borg']
SSH_ORIGINAL_COMMAND: None
Failed to prune archives!

Any Changes In Recent Updates That Can Cause This ?

**Comment by szaimen** [comment]
hello, can you run a backup check please and report back the result? Thanks!

**Comment by DrShivang** [comment]
Should i delete all the existing backups and redo the same ?

**Comment by szaimen** [comment]
> hello, can you run a backup check please and report back the result? Thanks!

**Comment by DrShivang** [comment]
Borg Check Integrity Logs :

> breaking the borg lock...
Local Exception
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 4690, in main
    exit_code = archiver.run(args)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 4622, in run
    return set_ec(func(args))
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 177, in wrapper
    return method(self, args, repository=repository, **kwargs)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 2236, in do_break_lock
    repository.break_lock()
  File "/usr/lib/python3/dist-packages/borg/repository.py", line 405, in break_lock
    Lock(os.path.join(self.path, 'lock')).break_lock()
  File "/usr/lib/python3/dist-packages/borg/locking.py", line 408, in break_lock
    self._lock.break_lock()
  File "/usr/lib/python3/dist-packages/borg/locking.py", line 204, in break_lock
    for name in os.listdir(self.path):
OSError: [Errno 74] Bad message: '/mnt/borgbackup/borg/lock.exclusive'
Platform: Linux 08edce88728c 5.10.103-v8 #1 SMP PREEMPT Mon Mar 28 11:44:31 UTC 2022 aarch64
Linux: Unknown Linux  
Borg: 1.1.16  Python: CPython 3.9.2 msgpack: 0.5.6.+borg1
PID: 16  CWD: /
sys.argv: ['/usr/bin/borg', 'break-lock', '/mnt/borgbackup/borg']
SSH_ORIGINAL_COMMAND: None
Could not break the borg lock!

**Comment by szaimen** [comment]
so it seems like maybe your drive is corrupt somehow...

**Comment by szaimen** [comment]
can you post the output of sudo cat /etc/mtab and grep for the drive where the backup is lying on? If it is mounted in read only mode there is probably something up with your drive...

**Comment by DrShivang** [comment]
cat /etc/mtab | grep /dev/sd
/dev/sda1 /mnt/data/supervisor/media/My_Passport ext4 rw,relatime 0 0

**Comment by szaimen** [comment]
okay, do I understand correctly that the backups get created in a subdir of `/mnt/data/supervisor/media/My_Passport`, yes?

**Comment by DrShivang** [comment]
> okay, do I understand correctly that the backups get created in a subdir of `/mnt/data/supervisor/media/My_Passport`, yes?

Yes

**Comment by szaimen** [comment]
Please try to delete the `lock.roster` in the backup directory manually and try to run the backup check again

**Comment by DrShivang** [comment]
is lock.exclusive the same thing ?

> Please try to delete the `lock.roster` in the backup directory manually and try to run the backup check again

**Comment by szaimen** [comment]
> is lock.exclusive the same thing ?

afaik, yes

**Comment by szaimen** [comment]
What was the result of the check?

**Comment by DrShivang** [comment]
> What was the result of the check?

Hi, Sorry for the delayed response, Integrity checks took a lot of time because of backup size and I needed to try a few times before commenting anything.

I did delete the file as mentioned above, and It doesn't seem to help.

Borg Directory :
>  borg ls -la
total 2596
drwxr-xr-x    4 root     root          4096 May  4 07:30 .
drwxr-xr-x    3 xfs      root          4096 Mar 23 20:24 ..
-rw-------    1 root     root            73 Mar 23 20:26 README
-rw-------    1 root     root           965 Mar 23 20:26 config
drwx------    3 root     root          4096 Mar 23 20:26 data
-rw-------    1 root     root           209 May  3 17:22 hints.68
-rw-------    1 root     root       2621498 May  3 17:22 index.68
-rw-------    1 root     root           190 May  3 17:22 integrity.68
-rw-------    1 root     root            16 May  3 17:22 nonce

Latest Borg Backup Log :
 It Doesn't Even Run Now .
 I have deleted 2 lock files from directory
1) lock.roster and 2) lock.exclusive
> Breaking the borg lock...
Local Exception
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 4690, in main
    exit_code = archiver.run(args)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 4622, in run
    return set_ec(func(args))
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 177, in wrapper
    return method(self, args, repository=repository, **kwargs)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 2236, in do_break_lock
    repository.break_lock()
  File "/usr/lib/python3/dist-packages/borg/repository.py", line 405, in break_lock
    Lock(os.path.join(self.path, 'lock')).break_lock()
  File "/usr/lib/python3/dist-packages/borg/locking.py", line 408, in break_lock
    self._lock.break_lock()
  File "/usr/lib/python3/dist-packages/borg/locking.py", line 204, in break_lock
    for name in os.listdir(self.path):
OSError: [Errno 74] Bad message: '/mnt/borgbackup/borg/lock.exclusive'
Platform: Linux 6a9b92ad0b3c 5.10.103-v8 #1 SMP PREEMPT Mon Mar 28 11:44:31 UTC 2022 aarch64
Linux: Unknown Linux  
Borg: 1.1.16  Python: CPython 3.9.2 msgpack: 0.5.6.+borg1
PID: 17  CWD: /
sys.argv: ['/usr/bin/borg', 'break-lock', '/mnt/borgbackup/borg']
SSH_ORIGINAL_COMMAND: None
Could not break the borg lock!

**Comment by DrShivang** [comment]
I just noticed the logs are the same irrespective of the fact that I run backup integrity check or run backup, both output the same log.
maybe dealing with this ghostly lock file is what both scripts do at the beginning.

Logs from Backup Integrity Check Yesterday : https://github.com/nextcloud/all-in-one/issues/554#issuecomment-1116036810
Logs from Attempted Backup Run Today : https://github.com/nextcloud/all-in-one/issues/554#issuecomment-1117265909

**Comment by szaimen** [comment]
Hm... honestly I don't have much ideas what could have triggered this issue. Also if that would be a common issue I suppose much more people would report the same issue...

Probably the easiest is to start over with a new backup archive.

I think for that you need to clear the borg directory and delete this file /var/lib/docker/volumes/nextcloud_aio_mastercontainer/_data/data/borg.config
and you should probably also delete the nextcloud_aio_backup_cache volume

**Comment by szaimen** [comment]
I'll leave this open in case this happens also for other people...

**Comment by szaimen** [comment, marked as answer]
I googled a bit more about these issues and everything points towards a faulty drive... 🤷‍♂️

It could if this is an external drive also be the usb cable that connect your your server... 
I guess running a smartctl -a /dev/sdx for that drive would also be interesting...

**Comment by DrShivang** [comment]
Yep, as nothing else seems to be the cause, we should hunt for the most common ones first.. I'll re-run diagnostics on drive again to see if I get anything this time, else I'll follow the advice: https://github.com/nextcloud/all-in-one/issues/554#issuecomment-1117281446 to reinitiate the borg backup from scratch.

Smartctl, I'll install and see if I get anything.

Thanks

**Comment by szaimen** [comment]
Hello @DrShivang were you able to find the root cause of this? And did it happen again?

**Comment by DrShivang** [comment]
Hi,

Due to lack of any supporting evidence and as per the google search i came
down to most likely cause : Disk Drive
Although the disk drive is new and didn't show any errors during file
system check still i removed the drive, deleted the borg directory and will
setup the backup directory again (awaited as i need some time because am
also changing the Nextcloud domain).

I don't think the error is gonna pop-up again.
I'll update incase i still face any issues


Thanks

**Comment by szaimen** [comment]
Allright, thanks for the reply!

It might have been the cable as I already pointed out but I guess you already checked that...

**Comment by DrShivang** [comment]
Yup, any of the hardware could be the cause of the problem
Yes, as you told earlier, I have reassembled it all again.

**Comment by szaimen** [comment]
Converting to a discussion since this is not a bug on our side and also no doc ehnancement/exiating feature request.
